README.md
# tsid-python

A Python library for generating Time-Sorted Unique Identifiers (TSID) as defined in <https://github.com/f4b6a3/tsid-creator>.

This library is a port of the original Java code by [Fabio Lima](https://github.com/fabiolimace).

## Installation

```bash
pip install tsidpy
```

## What is a TSID?

The term TSID stands for (roughly) Time-Sorted ID. A TSID is a value that is formed by its creation time along with a random value.

It brings together ideas from [Twitter's Snowflake](https://github.com/twitter-archive/snowflake/tree/snowflake-2010) and [ULID Spec](https://github.com/ulid/spec).

In summary:

- Sorted by generation time.
- Can be stored as a 64-bit integer.
- Can be stored as a 13-char len string.
- String format is encoded to [Crockford's base32](https://www.crockford.com/base32.html).
- String format is URL safe, case insensitive and has no hyphens.
- Shorter than other unique identifiers, like [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier), [ULID](https://github.com/ulid/spec) and [KSUID](https://github.com/segmentio/ksuid).

## TSID Structure

A TSID has 2 components:

1. A time component (42 bits), consisting in the elapsed milliseconds since `2020-01-01 00:00:00 UTC` (this epoch can be configured)
2. A _random_ component (22 bits), containing 2 sub-parts:

    - A node identifier (can use 0 to 20 bits)
    - A counter (can use 2 to 22 bits)

    > Note: The counter length depends on the node identifier length.
    >
    > For example, if we use 10 bits for the node representation:
    >
    > - The counter is limited to 12 bits.
    > - The maximum node value is `2^10-1 = 1023`
    > - The maximum counter value is `2^12-1 = 4095`, so the maximum TSIDs that can be generated _per millisecond_ is `4096`.

This is the default TSID structure:

```text
                                            adjustable
                                           <---------->
|------------------------------------------|----------|------------|
       time (msecs since 2020-01-01)           node      counter
                42 bits                       10 bits    12 bits

- time:    2^42 = ~69 to ~139 years with adjustable epoch (see notes below)
- node:    up to 2^20 values with adjustable bits.
- counter: 2^2..2^22 with adjustable bits and randomized values every millisecond.
```

> Notes:
>
> - The time component can be used for ~69 years if stored in a `SIGNED 64-bit` integer field (41 usable bits) or ~139 years if stored in a `UNSIGNED 64-bit` integer field (42 usable bits).
> - By default, new TSID generators use 10 bits for the node identifier and 12 bits to the counter. It's possible to adjust the node identifier length to a value between 0 and 20.
> - The time component can be 1 ms or more ahead of the system time when necessary to maintain monotonicity and generation speed.

### Node identifier

The simplest way to avoid collisions is to make sure that each generator has an exclusive node ID.

The node ID can be passed to the `TSIDGenerator` constructor. If no node ID is passed, the generator will use a random value.

### Recommended readings

- [The best UUID type for a database Primary Key](https://vladmihalcea.com/uuid-database-primary-key/)
- [The primary key dilemma: ID vs UUID and some practical solutions](https://fillumina.wordpress.com/2023/02/06/the-primary-key-dilemma-id-vs-uuid-and-some-practical-solutions/)
- [Primary keys in the DB - what to use? ID vs UUID or is there something else?](https://www.linkedin.com/pulse/primary-keys-db-what-use-id-vs-uuid-something-else-lucas-persson)

Related with the [original library](https://github.com/f4b6a3/tsid-creator):

- [FAQ wiki page](https://github.com/f4b6a3/tsid-creator/wiki)
- [Javadocs](https://javadoc.io/doc/com.github.f4b6a3/tsid-creator)
- [How to not use TSID factories](https://fillumina.wordpress.com/2023/01/19/how-to-not-use-tsid-factories/)
- [The best way to generate a TSID entity identifier with JPA and Hibernate](https://vladmihalcea.com/tsid-identifier-jpa-hibernate/)

## Basic usage

Create a TSID:

```python
from tsidpy import TSID

tsid: TSID = TSID.create()
```

Create a TSID as an `int`:

```python
>>> TSID.create().number
432511671823499267
```

Create a TSID as a `str`:

```python
>>> str(TSID.create())
'0C04Q2BR40003'
```

Create a TSID as an hexadecimal `str`:

```python
>>> TSID.create().to_string('x')
'06009712f0400003'
```

> Note: TSID generators are [thread-safe](https://en.wikipedia.org/wiki/Thread_safety).

### TSID as int

The `TSID::number` property simply unwraps the internal `int` value of a TSID.

```python
>>> from tsidpy import TSID
>>> TSID.create(432511671823499267).number
432511671823499267
```

Sequence of TSIDs:

```text
38352658567418867
38352658567418868
38352658567418869
38352658567418870
38352658567418871
38352658567418872
38352658567418873
38352658567418874
38352658573940759 < millisecond changed
38352658573940760
38352658573940761
38352658573940762
38352658573940763
38352658573940764
38352658573940765
38352658573940766
         ^      ^ look
                                   
|--------|------|
   time   random
```

### TSID as str

The `TSID::to_string()` method encodes a TSID as a [Crockford's base 32](https://www.crockford.com/base32.html) string. The returned string is 13 characters long.

```python
>>> from tsidpy import TSID
>>> tsid: str = TSID.create().to_string()
'0C04Q2BR40004'
```

Or, alternatively:

```python
>>> tsid: str = str(TSID.create())
'0C04Q2BR40004'
```

Sequence of TSID strings:

```text
01226N0640J7K
01226N0640J7M
01226N0640J7N
01226N0640J7P
01226N0640J7Q
01226N0640J7R
01226N0640J7S
01226N0640J7T
01226N0693HDA < millisecond changed
01226N0693HDB
01226N0693HDC
01226N0693HDD
01226N0693HDE
01226N0693HDF
01226N0693HDG
01226N0693HDH
        ^   ^ look
                                   
|-------|---|
   time random
```

The string format can be useful for languages that store numbers in [IEEE 754 double-precision binary floating-point format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format), such as [Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number).

### More Examples

Create a TSID using the default generator:

```python
from tsidpy import TSID

tsid: TSID = TSID.create()
```

---

Create a TSID from a canonical string (13 chars):

```python
from tsidpy import TSID

tsid: TSID = TSID.from_string('0123456789ABC')
```

---

Convert a TSID into a canonical string in lower case:

```python
>>> tsid.to_string('s')
'0123456789abc'
```

---

Get the creation `timestamp` of a TSID:

```python
>>> tsid.timestamp
1680948418241.0  # datetime.datetime(2023, 4, 8, 12, 6, 58, 241000)
```

---

Encode a TSID to base-62:

```python
>>> tsid.to_string('z')
'0T5jFDIkmmy'
```

---

A `TSIDGenerator` that creates TSIDs similar to [Twitter Snowflakes](https://github.com/twitter-archive/snowflake):

- Twitter snowflakes use 10 bits for node id: 5 bits for datacenter ID (max 31) and 5 bits for worker ID (max 31)
- Epoch starts on `2010-11-04T01:42:54.657Z`
- Counter uses 12 bits and starts at `0` (max: 4095 values per millisecond)

```python
from tsidpy import TSID, TSIDGenerator

datacenter: int = 1
worker: int = 1
node: int = datacenter << 5 | worker
epoch: datetime = datetime.fromisoformat('2010-11-04T01:42:54.657Z')

twitter_generator: TSIDGenerator = TSIDGenerator(node=node, node_bits=10,
                                                 epoch=epoch.timestamp() * 1000,
                                                 random_fn=lambda n: 0)

# use the generator
tsid: TSID = twitter_generator.create()
```

---

A `TSIDGenerator` that creates TSIDs similar to [Discord Snowflakes](https://discord.com/developers/docs/reference#snowflakes):

- Discord snowflakes use 10 bits for node id: 5 bits for worker ID (max 31) and 5 bits for process ID (max 31)
- Epoch starts on `2015-01-01T00:00:00.000Z`
- Counter uses 12 bits and starts at a random value.

```python
from tsidpy import TSID, TSIDGenerator

worker: int = 1
process: int = 1
node: int = worker << 5 | process
epoch: datetime = datetime.fromisoformat("2015-01-01T00:00:00.000Z")

discord_generator: TSIDGenerator = TSIDGenerator(node=node, node_bits=10,
                                                 epoch=epoch.timestamp() * 1000)

# use the generator
tsid: TSID = discord_generator.create()
```

---

Make `TSID.create()` to use the previous Discord generator:

```python

TSID.set_default_generator(discord_generator)

# at this point, you can use the default TSID.create()
tsid: TSID = TSID.create()

# or the generator
tsid: TSID = discord_generator.create()
```

---

### A note about node id and node bits

When creating a `TSIDGenerator`, remember you can't use a node id greater than `2^node_bits - 1`. For example, if you need to use a node id greater than 7, you need to use more than 3 bits for the node id:

```python
from tsidpy import TSIDGenerator

gen0 = TSIDGenerator(node=0, node_bis=3)  # ok
gen1 = TSIDGenerator(node=1, node_bis=3)  # ok
...
gen7 = TSIDGenerator(node=7, node_bis=3)  # ok

# error: can't represent 8 with 3 bits
gen8 = TSIDGenerator(node=8, node_bis=3)
```

## Other ports, forks and OSS

### Ports and forks

| Language | Name |
| -------- | ---- |
| Java (by original author)     | [f4b6a3/tsid-creator](https://github.com/f4b6a3/tsid-creator) |
| Java     | [vladmihalcea/hypersistence-tsid](https://github.com/vladmihalcea/hypersistence-tsid) |
| .NET     | [kgkoutis/TSID.Creator.NET](https://github.com/kgkoutis/TSID.Creator.NET) |
| PHP      | [odan/tsid](https://github.com/odan/tsid) |

### OSS

| Language | Name |
| -------- | ---- |
| Java     | [fillumina/id-encryptor](https://github.com/fillumina/id-encryptor) |
| .NET     | [ullmark/hashids.net](https://github.com/ullmark/hashids.net) |

## License

This library is Open Source software released under the [MIT license](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.1.5] - 2023-11-28

### Added

- Make `TSID` values hashable.

## [1.1.4] - 2023-11-09

### Added

- `datetime` property.

## [1.1.3] - 2023-10-09

### Fixed

- Honor Python 3.7 syntax .

## [1.1.2] - 2023-09-13

### Fixed

- Properly exported symbols as per https://github.com/microsoft/pylance-release/issues/2953#issuecomment-1168408943.

## [1.1.1] - 2023-08-18

### Added

- Package type hints (thanks to https://github.com/kasium) (#2).
- `CHANGELOG.md` file and proper semver schema adhesion (#3).


perf_test.py
import argparse
import threading
import typing as t
from uuid import uuid4

from tsid import TSID

cache: set[int] = set()


def tsid_create() -> None:
    _ = TSID.create().number


def tsid_create_str() -> None:
    _ = str(TSID.create())


def uuid_create() -> None:
    _ = uuid4()


def uuid_create_str() -> None:
    _ = str(uuid4())


class CreationThread(threading.Thread):
    def __init__(self, loops: int, perf_fn: t.Callable[[], None]):
        self.loops: int = loops
        self.perf_fn: t.Callable[[], None] = perf_fn
        super().__init__()

    def run(self):
        for _ in range(self.loops):
            self.perf_fn()


def main(mode: str, loops: int, thread_count: int) -> None:
    fn: t.Callable[[], None]

    match mode:
        case 'tsid':
            fn = tsid_create
        case 'tsid_str':
            fn = tsid_create_str
        case 'uuid':
            fn = uuid_create
        case 'uuid_str':
            fn = uuid_create_str
        case _:
            raise ValueError('Invalid mode')

    threads: list[CreationThread] = []
    for _ in range(thread_count):
        thread: CreationThread = CreationThread(loops, fn)
        thread.start()
        threads.append(thread)

    print('Waiting for threads to finish...')
    for thread in threads:
        thread.join()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('mode', type=str, help='Which mode to run in',
                        choices=['tsid', 'tsid_str', 'uuid', 'uuid_str'])
    parser.add_argument('loops', type=int, default=1000000,
                        help='Number of loops per thread')
    parser.add_argument('threads', type=int, default=4,
                        help='Number of threads to use')

    args = parser.parse_args()

    main(args.mode, args.loops, args.threads)


collision_test.py
import threading
from datetime import datetime
from tsid.tsid import TSID

cache: set[int] = set()

class CreationThread(threading.Thread):
    def __init__(self):
        self.count: int = 0
        super().__init__()

    def run(self):
        while True:
            t: TSID = TSID.create()
            if t.number in cache:
                print(f"Collision ({self.count}): {t.number} {t.random} {t.datetime} {datetime.now()}")
            else:
                cache.add(t.number)

            self.count += 1

            if self.count % 1000000 == 0:
                print(f"Thread {self.name}, count: {self.count}")

print("Testing collision. Press Ctrl+C to stop.")

threads: list[CreationThread] = []

for i in range(8):
    t: CreationThread = CreationThread()
    t.start()
    threads.append(t)

for t in threads:
    t.join()


tests/perf_test.py
import argparse
import threading
import typing as t
from uuid import uuid4

from tsidpy import TSID


DEFAULT_LOOPS = 100000
DEFAULT_THREADS = 4


def tsid_create() -> None:
    _ = TSID.create().number


def tsid_create_str() -> None:
    _ = str(TSID.create())


def uuid_create() -> None:
    _ = uuid4()


def uuid_create_str() -> None:
    _ = str(uuid4())


class CreationThread(threading.Thread):
    def __init__(self, loops: int, perf_fn: t.Callable[[], None]):
        self.loops: int = loops
        self.perf_fn: t.Callable[[], None] = perf_fn
        super().__init__()

    def run(self):
        for _ in range(self.loops):
            self.perf_fn()


def main(mode: str, loops: int, thread_count: int) -> None:
    fn: t.Callable[[], None]

    match mode:
        case 'tsid':
            fn = tsid_create
        case 'tsid_str':
            fn = tsid_create_str
        case 'uuid':
            fn = uuid_create
        case 'uuid_str':
            fn = uuid_create_str
        case _:
            raise ValueError('Invalid mode')

    threads: list[CreationThread] = [CreationThread(loops, fn)
                                     for _ in range(thread_count)]

    for thread in threads:
        thread.start()

    print('Waiting for threads to finish...')
    for thread in threads:
        thread.join()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="""
This program tests the performance of the TSID class and its string
serialization methods. It also allows to compare it with the performance of
uuid4() and str(uuid4()).
""")
    parser.add_argument('mode', type=str, help='Which mode to run in',
                        choices=['tsid', 'tsid_str', 'uuid', 'uuid_str'])
    parser.add_argument('--loops', type=int, default=DEFAULT_LOOPS,
                        required=False, help='Number of loops per thread')
    parser.add_argument('--threads', type=int, default=DEFAULT_THREADS,
                        required=False, help='Number of threads to use')

    args = parser.parse_args()

    main(args.mode, args.loops, args.threads)


src/tsidpy/tsid.py

import functools
import random
import threading
import time
import typing as t

from datetime import datetime

from .basen import decode, encode

TSID_BYTES: int = 8
TSID_CHARS: int = 13
TSID_HEXCHARS: int = 16

_EPOCH_ISO: str = '2020-01-01 00:00:00+00:00'
TSID_EPOCH: float = datetime.fromisoformat(_EPOCH_ISO).timestamp() * 1000

RANDOM_BITS: int = 22
RANDOM_MASK: int = 0x3fffff

TSID_DEFAULT_NODE_BITS = 10

ALPHABET: str = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'
ALPHABET_VALUES: t.List[int] = [-1 for _ in range(128)]


def __set_alphabet_values(chars: str, start: int) -> None:
    for i, c in enumerate(chars):
        ALPHABET_VALUES[ord(c)] = start + i
        ALPHABET_VALUES[ord(c.upper())] = start + i


__set_alphabet_values('0123456789', 0)
__set_alphabet_values('abcdefghjkmnpqrstvwxyz', 0xa)
__set_alphabet_values('oi', 0)
__set_alphabet_values('l', 1)


_default_generator: 'TSIDGenerator'


@functools.total_ordering
class TSID:
    """A value object that represents a Time-Sorted Unique Identifier (TSID).

       TSID is a 64-bit value that has 2 components:

       - Time component (42 bits): a number of milliseconds since 2020-01-01
         (Unix epoch).
       - Random component (22 bits): a sequence of random bits generated by a
         secure random generator.

       The Random component has 2 sub-parts:
       - Node (0 to 20 bits): a number used to identify the machine or node.
       - Counter (2 to 22 bits): a randomly generated number that is
         incremented whenever the time component is repeated.

       The random component layout depend on the node bits. If the node bits
       are 10, the counter bits are limited to 12. In this example, the
       maximum node value is `2^10-1 = 1023` and the maximum counter value is
       `2^12-1 = 4093`. So the maximum TSIDs that can be generated per
       millisecond per node is `4096`.

       Instances of this class are immutable.

       See [Snowflake ID](https://en.wikipedia.org/wiki/Snowflake_ID).
    """

    def __init__(self, number: int) -> None:
        """
        >>> TSID(0x10000000000000000).number == 0
        True
        """
        self.__number: int = number & 0xffffffffffffffff  # 64-bit
        self._epoch: float = TSID_EPOCH
    
    def __hash__(self) -> int:
        """
        >>> hash(TSID(0))
        0
        >>> hash(TSID(1))
        1
        >>> hash(TSID(0xffffffffffffffff)) == hash(0xffffffffffffffff)
        True
        >>> d = dict()
        >>> d[TSID(0)] = 0
        >>> d[TSID(1)] = 1
        >>> d[TSID(0xffffffffffffffff)] = "big"
        >>> d[TSID(0)]
        0
        >>> d[TSID(1)]
        1
        >>> d[TSID(0xffffffffffffffff)]
        'big'
        """
        return self.__number.__hash__()

    @property
    def timestamp(self) -> float:
        """Returns the timestamp component of the TSID.
           The timestamp is the number of milliseconds elapsed
           since the TSID epoch.

        >>> TSID(0).timestamp == TSID_EPOCH
        True
        >>> TSID.create().timestamp - time.time() * 1000 < 1
        True
        >>> TSID(1 << RANDOM_BITS).timestamp == TSID_EPOCH + 1
        True
        """
        return self._epoch + (self.__number >> RANDOM_BITS)

    @property
    def datetime(self) -> datetime:
        """Datetime corresponding with the tiimestamp component
           of the TSID.

        >>> TSID(0).datetime == datetime.fromtimestamp(TSID_EPOCH / 1000)
        True
        """
        return datetime.fromtimestamp(self.timestamp / 1000)

    @property
    def random(self) -> int:
        """Returns the random component of the TSID.
           This compomnent contains the node and counter
           bits.

        >>> TSID(0).random == 0
        True
        >>> TSID(1).random == 1
        True
        >>> TSID((0xffffffff << RANDOM_BITS) + 255).random == 255
        True
        """
        return self.__number & RANDOM_MASK

    @property
    def number(self) -> int:
        """Converts the TSID into a 64-bit number.
           This simply unwraps the internal value

        >>> TSID(0xffff0000000000000000).number == 0
        True
        >>> TSID(0x0000000000000000).number == 0
        True
        >>> TSID(0xffff).number == 0xffff
        True
        """
        return self.__number

    def __lt__(self, other: object) -> bool:
        """
        >>> TSID(0) < TSID(1)
        True
        >>> TSID(1) > TSID(0)
        True
        >>> TSID(0) < 1000
        False
        """
        if isinstance(other, TSID):
            return self.__number < other.__number
        return False

    def __eq__(self, other: object) -> bool:
        """
        >>> TSID(0) == TSID(0)
        True
        >>> TSID(0) != 0
        True
        >>> TSID(0) != '0'
        True
        """
        if isinstance(other, TSID):
            return self.__number == other.__number
        return False

    def __repr__(self) -> str:
        return self._to_canonical_string()

    def __str__(self) -> str:
        return self._to_canonical_string()

    def to_bytes(self) -> bytes:
        r"""Converts the TSID into a byte array.

        >>> t = TSID(1)
        >>> t.to_bytes() == b'\x00\x00\x00\x00\x00\x00\x00\x01'
        True

        >>> t = TSID(0xfabada)
        >>> t.to_bytes() == b'\x00\x00\x00\x00\x00\xfa\xba\xda'
        True

        >>> t = TSID(0xffcafefabadabeef)
        >>> t.to_bytes() == b'\xff\xca\xfe\xfa\xba\xda\xbe\xef'
        True

        >>> t = TSID(0xffffffcafefabadabeef)
        >>> t .to_bytes() == b'\xff\xca\xfe\xfa\xba\xda\xbe\xef'
        True
        """
        return self.__number.to_bytes(TSID_BYTES, byteorder='big',
                                      signed=False)

    def to_string(self, fmt: str = 'S') -> str:
        """Converts the TSID into a string.

           Supports the following formats:
            - `S`: canonical string in upper case.
            - `s`: canonical string in lower case.
            - `X`: hexadecimal in upper case.
            - `x`: hexadecimal in lower case.
            - `d`: base-10.
            - `z`: base-62.

        >>> t = TSID.from_string('0AWE5HZP3SKTK')
        >>> t.to_string('S')
        '0AWE5HZP3SKTK'
        >>> t.to_string('x')
        '0571c58fec3ccf53'

        >>> t = TSID.from_string('0AXFXR5W7VBX0')
        >>> t.to_string('X')
        '0575FDC1787DAFA0'

        >>> t = TSID(0xffffffffffffffff)
        >>> t.to_string()
        'FZZZZZZZZZZZZ'

        >>> t = TSID(0x000000000000000a)
        >>> t.to_string()
        '000000000000A'

        >>> t = TSID(437283649808777971)
        >>> t.to_string('d')
        '437283649808777971'
        >>> t.to_string('z')
        'WIlLsHwljH'
        """

        result: str

        if fmt == 'S':  # canonical string in upper case
            result = self._to_canonical_string()
        elif fmt == 's':  # canonical string in lower case
            result = self._to_canonical_string().lower()
        elif fmt == 'X':  # hexadecimal in upper case
            result = encode(self.number, 16, min_length=TSID_BYTES*2)
        elif fmt == 'x':  # hexadecimal in lower case
            result = encode(self.number, 16, min_length=TSID_BYTES*2)
            result = result.lower()
        elif fmt == 'd':  # base-10
            result = encode(self.number, 10)
        elif fmt == 'z':  # base-62
            result = encode(self.number, 62)
        else:
            raise ValueError(f"Invalid format: '{fmt}'")

        return result

    def _to_canonical_string(self) -> str:
        return ''.join(ALPHABET[self.__number >> i & 0x1f]
                       for i in range(60, -5, -5))

    @staticmethod
    def create() -> 'TSID':
        """Returns a new TSID.

           This static method is a quick alternative to
           `TSIDGenerator::create()`.

           It can generate up to `2^22` (`4,194,304`) TSIDs per millisecond.
           It can be useful, for example, for logging.

           Security-sensitive applications that require a cryptographically
           secure pseudo-random generator should use `TSIDGenerator::create()`.

           > Note this method is not thread safe by default. It's the
           TSIDGenerator responsibility to ensure thread safety.

        >>> a = TSID.create()
        >>> b = TSID.create()
        >>> c = TSID.create()
        >>> a.number < b.number < c.number
        True
        """
        return _default_generator.create()

    @staticmethod
    def from_bytes(bytes: bytes) -> 'TSID':
        r"""Converts a byte array into a TSID.

        >>> TSID.from_bytes(b'\x00\x00\x00\x00\x00\x00\x00\x00') == TSID(0)
        True
        >>> TSID.from_bytes(b'\x00\x00\x00\x00\x00\x00\x00\x01') == TSID(1)
        True
        >>> TSID.from_bytes(b'\x00\x00\x00\x00\x00\x00\x00\x02') == TSID(2)
        True
        >>> TSID.from_bytes(b'\x00\x00\x00\x00\x00\x00\x00\x0b') == TSID(11)
        True
        """
        if len(bytes) != TSID_BYTES:
            raise ValueError(f'Invalid TSID bytes (len={len(bytes)} bytes, '
                             f'expected {TSID_BYTES})')

        number: int = int.from_bytes(bytes, byteorder='big', signed=False)
        return TSID(number)

    @staticmethod
    def from_string(value: str, fmt: str = 'S') -> 'TSID':
        """Converts a string into a TSID.

           Supports the following formats:
            - `S`: canonical string in upper case.
            - `s`: canonical string in lower case.
            - `X`: hexadecimal in upper case.
            - `x`: hexadecimal in lower case.
            - `d`: base-10.
            - `z`: base-62.

        >>> t1 = TSID.from_string('0575FDC1787DAFA0', 'X')
        >>> t2 = TSID.from_string('0AXFXR5W7VBX0')
        >>> t1.number == t2.number
        True
        """
        number: int

        if fmt == 'S' or fmt == 's':
            if len(value) != TSID_CHARS:
                raise ValueError(f'Invalid TSID string: '
                                    f'(len={len(value)} chars, '
                                    f'but expected {TSID_CHARS})')
            number = sum(ALPHABET_VALUES[ord(value[i])] << h
                            for i, h in enumerate(range(60, -5, -5), 0))
        elif fmt == 'X':  # hexadecimal in upper case
            if len(value) != TSID_HEXCHARS:
                raise ValueError(f'Invalid TSID string: '
                                    f'(len={len(value)} chars, '
                                    f'but expected {TSID_HEXCHARS})')
            number = decode(value, 16)
        elif fmt == 'x':  # hexadecimal in lower case
            if len(value) != TSID_HEXCHARS:
                raise ValueError(f'Invalid TSID string: '
                                    f'(len={len(value)} chars, '
                                    f'but expected {TSID_HEXCHARS})')
            number = decode(value.upper(), 16)
        elif fmt == 'd':  # base-10
            number = decode(value, 10)
        elif fmt == 'z':  # base-62
            number = decode(value, 62)
        else:
            raise ValueError(f"Invalid format: '{fmt}'")

        return TSID(number)

    @staticmethod
    def set_default_generator(generator: 'TSIDGenerator') -> None:
        """Sets the default TSID generator.

        >>> generator: TSIDGenerator = TSIDGenerator(node=1)
        >>> TSID.set_default_generator(generator)
        """
        global _default_generator
        _default_generator = generator


class TSIDGenerator:
    def __init__(
        self,
        node: t.Optional[int] = None,
        node_bits: int = TSID_DEFAULT_NODE_BITS,
        epoch: float = TSID_EPOCH,
        random_fn: t.Optional[t.Callable[[int], int]] = None
    ) -> None:
        """Creates a new TSID generator.

        Args:
        - `node` int | None: Node identifier. Defaults to a random integer
                             within the range of `node_bits`.
        - `node_bits` int:   Number of bytes used to repesent the node id.
                             Defaults to `TSID_DEFAULT_NODE_BITS`.
        - `epoch` int:       Epoch start in milliseconds. Defaults to
                             `TSID_EPOCH`.
        - `random_fn`:       Function to use to randomize the counter.
                             Must return a n-bit integer. If None, the
                             stdlib `random.getrandbits()` is used.

        >>> generator = TSIDGenerator(node=1, node_bits=21)
        Traceback (most recent call last):
        ...
        ValueError: Invalid node_bits: 21

        >>> generator = TSIDGenerator(node=1, node_bits=0)
        Traceback (most recent call last):
        ...
        ValueError: Node 1 too large for node_bits==0

        >>> generator = TSIDGenerator(node=1, node_bits=-1)
        Traceback (most recent call last):
        ...
        ValueError: Invalid node_bits: -1

        >>> generator = TSIDGenerator(node=1, node_bits=1)
        >>> generator.node
        1
        """
        if node is not None and node < 0:
            raise ValueError(f'Invalid node: {node}')

        if node_bits < 0 or node_bits > 20:
            raise ValueError(f'Invalid node_bits: {node_bits}')

        self.random_fn = random_fn or random.getrandbits

        self.node: int
        if node is None:
            self.node = random.getrandbits(20) >> (20 - node_bits)
        else:
            self.node = node

        if self.node >> node_bits > 0:
            raise ValueError(f'Node {self.node} too large for '
                             f'node_bits=={node_bits}')

        self._epoch: float = epoch
        self._node_bits: int = node_bits
        self._counter_bits: int = RANDOM_BITS - node_bits
        self.counter: int = self.random_fn(self._counter_bits)

        self._millis: float = time.time() * 1000
        self._counter_mask: int = RANDOM_MASK >> node_bits
        self._node_mask: int = RANDOM_MASK >> self._counter_bits

        self._lock = threading.Lock()

    def create(self) -> TSID:
        """
        >>> ### Test node extraction ------------------------------
        >>> tc = TSIDGenerator(node=255, node_bits=8)
        >>> t = tc.create()
        >>> ((t.number & RANDOM_MASK) >> tc._counter_bits) == 255
        True
        >>> tc = TSIDGenerator(node=64, node_bits=8)
        >>> t = tc.create()
        >>> ((t.number & RANDOM_MASK) >> tc._counter_bits) == 64
        True
        >>> tc = TSIDGenerator(node=512, node_bits=10)
        >>> t = tc.create()
        >>> ((t.number & RANDOM_MASK) >> tc._counter_bits) == 512
        True

        >>> ### Test counter extraction ------------------------------
        >>> tc = TSIDGenerator(node=64, node_bits=8, random_fn=lambda n: 0)
        >>> t = tc.create()
        >>> t.number & tc._counter_mask == 1
        True
        >>> t = tc.create()
        >>> t.number & tc._counter_mask == 2
        True

        >>> ### Test random extraction ------------------------------
        >>> tc = TSIDGenerator(node=0, node_bits=0, random_fn=lambda n: 0)
        >>> t = tc.create()
        >>> t.random == 1
        True
        >>> t = tc.create()
        >>> t.random == 2
        True

        >>> ### Test timestamp extraction ------------------------------
        >>> t = tc.create()
        >>> t.timestamp - (time.time() * 1000) <= 1
        True

        >>> ### Test custom epoch ------------------------------
        >>> from datetime import datetime
        >>> UNIX_EPOCH = datetime.fromisoformat('1970-01-01 00:00:00+00:00')
        >>> EPOCH_MS = UNIX_EPOCH.timestamp() * 1000
        >>> now_without_unix_epoch: float = time.time() * 1000 - EPOCH_MS
        >>> tc = TSIDGenerator(node=0, epoch=0)
        >>> t = tc.create()
        >>> t.timestamp - now_without_unix_epoch <= 1
        True
        """
        with self._lock:
            current_millis: float = time.time() * 1000

            # If same millisecond, increment counter
            if current_millis - self._millis < 1:
                self.counter += 1

                # If the counter overflows, go to the next millisecond
                if self.counter >> self._counter_bits != 0:
                    self._millis += 1
                    self.counter = 0
            else:
                self._millis = current_millis
                rnd: int = self.random_fn(self._counter_bits)
                self.counter = rnd & self._counter_mask

            millis = int(self._millis - self._epoch) << RANDOM_BITS
            node = (self.node & self._node_mask) << self._counter_bits
            counter = self.counter & self._counter_mask

            result = TSID(millis + node + counter)
            result._epoch = self._epoch
            return result


_default_generator = TSIDGenerator(node_bits=0)
